### PR TITLE
WIP: Reimplement dump and implment restore command

### PR DESCRIPTION
**Pull request by lunny**
This PR would like to replace #1637. It reimplemented `dump` command and old dump file will not be used anymore. And also added a restore command that could restore dumped zip file to some machine. You have to use the same version Gitea with `dump` when `restore`.

The dump databases will be stored as `<table_name>.yml` in one directory. This format is expected to be compatible with Gitea's test fixtures. So that it's easy to export test data from a real system.

Currently it's blocked because yaml package didn't support Gonic name mapping, so I have forked yml package to https://github.com/go-gitea/yaml. Before that package supports that feature, this PR is not really work.

### CODE REVIEW

File: cmd/dump.go [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // Copyright 2014 The Gogs Authors. All rights reserved.
-// Copyright 2016 The Gitea Authors. All rights reserved.
+// Copyright 2017 The Gitea Authors. All rights reserved.
```
Comment (bkcsoft) [inline cmd/dump.go]: Don't change the copyright year ;)

**Review by bkcsoft — CHANGES_REQUESTED** [review]
I would keep the `gitea dump` command for a while and create a `gitea backup` command instead.

File: cmd/dump.go (hunk 1 above)
Comment (lunny) [reply cmd/dump.go]: But in fact all contents changed.

Comment (bkcsoft) [reply cmd/dump.go]: No it has not 🤔

**Comment by elvarb** [comment]
Any ETA on this?

**Review by bkcsoft — APPROVED** [review]

File: models/models.go [hunk 2]
```diff
@@ -347,3 +350,80 @@ func DumpDatabase(filePath string, dbType string) error {
...
+func restoreTableFixtures(bean interface{}, dirPath string) error {
+	table := x.TableInfo(bean)
+	data, err := ioutil.ReadFile(filepath.Join(dirPath, table.Name+".yml"))
+	if err != nil {
+		return err
+	}
+
+	var bufferSize = 100
+	v := reflect.MakeSlice(table.Type, 0, bufferSize)
+	// BLOCK: need yaml support gonic name mapper
```
Comment (techknowlogick) [inline models/models.go]: There have been 61 commits to `go-yaml/yaml` since it was forked into the `go-gitea` org. Is this still blocked?

Comment (lunny) [reply models/models.go]: I will check that.

File: cmd/restore.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
```
Comment (techknowlogick) [inline cmd/restore.go:1]: As this file is entirely new, I know you wrote it in 2017, however it is likely to get merged in 2018 should this date be updated to 2018?

**Comment by techknowlogick** [comment]
@lunny It looks like you've made some recent commits to this. Can the `blocked` label be removed?

File: cmd/restore.go (hunk 3 above)
Comment (lunny) [reply cmd/restore.go:1]: The copyright year should be the time first created I think.

**Comment by lunny** [comment]
@techknowlogick It still cannot work fine when restore.

**Comment by sphrak** [comment]
Is there a possibility this will be merged into 1.5.0?

**Comment by lunny** [comment]
@sphrak it's impossible since v1.5 has been freezed. I would like this could be in v1.6.

**Comment by jlelse** [comment]
Any updates on this? I would like to migrate from SQLite to Postgres and therefor need this feature.

**Comment by sphrak** [comment]
Any chance this will be part of 1.7.0?

**Comment by cryptix** [comment]
@lunny I'd love to try and test this but can share a little light on the `go-yaml/yaml` fork situation?

**Comment by lunny** [comment]
@cryptix you are welcome we need yaml support Gonic name mapping, maybe you could send a PR to yaml at first.

**Comment by dodedodo** [comment]
Any updates on this? Like jlelse I'm trying to do a database migration, and this should make my life so much easier.

**Comment by dev-dsp** [comment]
Any updates on this? I would like to migrate from SQLite to Postgres and therefore need this feature.

**Comment by dev-dsp** [comment]
For those who are curious about how to migrate data from sqlite to postgresql, see this link: https://github.com/go-gitea/gitea/issues/5651#issuecomment-661953408

**Comment by singuliere** [comment]
It looks like this PR is still relevant because the [restore process still is manual](https://docs.gitea.io/en-us/backup-and-restore/#restore-command-restore).

**Comment by yardenshoham** [comment]
This is in progress for a while now, I'm closing it. Please reopen when it's ready for review.

**Pull request closed without merging**